### PR TITLE
feat(migrate): simple db seed

### PR DESCRIPTION
**Pull request by Jolg42**
Seed location supported:
- `prisma/seed.js`
- `prisma/seed.ts`
- `prisma/seed.sh`
- `prisma/seed.go`

```
# Usage
prisma db seed --preview-feature

# No seed (could probably use a link to the docs later for an example)
Error: No seed file found.
Create a `seed.ts` or `.js` or `.sh` or `.go` file in the prisma folder.

# More than 1 seed
Error: More than 1 seed file was found in `prisma`.
This command only supports 1 `seed.ts` or `.js` or `.sh` or `.go` file.

# prisma/seed.js with a "Hello from seed.js"
Running `node seed.js` ...
Hello from seed.js

🌱  Your database is now seeded.

# prisma/seed.js with a "Hello from seed.js"
Running `node seed.js` ...
Hello from seed.js

🌱  Your database is now seeded.

# prisma/seed.ts with a "Hello from seed.js"
Running `ts-node seed.ts` ...
Hello from seed.js

🌱  Your database is now seeded.
```

Also in migrate reset, at the end of reset (after drop, create, apply, generate if not skipped) it will check if seed scripts are present.

If 0 seed script, nothing happens
If 1 seed script we try to run it
If +1 scripts error that it only supports one like ⬆️ 
```
prisma migrate reset

# Output
Prisma schema loaded from prisma/schema.prisma
Datasource "db": SQLite database "dev.db" at "file:./dev.db"


✔ Are you sure you want to reset your database? All data will be lost. … yes
Database reset successful
Running `ts-node seed.ts` ...
Hello from seed.js

```

### CODE REVIEW

File: src/packages/cli/src/bin.ts [hunk 1]
```diff
@@ -153,6 +154,7 @@ async function main(): Promise<number> {
         pull: Introspect.new(),
         push: DbPush.new(),
         drop: DbDrop.new(),
+        seed: DbDrop.new(),
```
Comment (tomhoule) [inline src/packages/cli/src/bin.ts:157]: Is the `DbDrop` here intentional?

Comment (Jolg42) [reply src/packages/cli/src/bin.ts:157]: Haha nice one!

File: src/packages/migrate/src/utils/seed.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    ts: fs.existsSync(tsFilePath) ? tsFilePath : null,
+    sh: fs.existsSync(shFilePath) ? shFilePath : null,
+    go: fs.existsSync(goFilePath) ? goFilePath : null,
+  }
+
+  detected.numberOfSeedFiles =
+    Number(!!detected.js) +
+    Number(!!detected.ts) +
+    Number(!!detected.sh) +
+    Number(!!detected.go)
```
Comment (steebchen) [inline src/packages/migrate/src/utils/seed.ts:26]: how about:

```suggestion
  const detected = {
    seedPath,
    numberOfSeedFiles: 0,
  }
  
  const paths = ['js', 'ts', 'sh', 'go']

  for (const path of paths) {
    if (!fs.existsSync(jsFilePath) {
      continue
    }
    const fullPath = seedPath + path
    detected[path] = fullPath
    detected.numberOfSeedFiles++
  }
```

File: src/packages/migrate/src/utils/seed.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  } else if (detected.numberOfSeedFiles > 1) {
+    throw new Error(
+      `More than 1 seed file was found in \`${path.relative(
+        process.cwd(),
+        path.dirname(detected.seedPath),
+      )}\`.
+This command only supports 1 \`seed.ts\` or \`.js\` or \`.sh\` or \`.go\` file.`,
+    )
+  } else {
+    if (detected.js) {
```
Comment (steebchen) [inline src/packages/migrate/src/utils/seed.ts:46]: use switch case? :D

**Review by steebchen — APPROVED** [review]
looks like a good default 👍

File: src/packages/migrate/src/utils/seed.ts (hunk 3 above)
Comment (Jolg42) [reply src/packages/migrate/src/utils/seed.ts:46]: Personally the switch cases don't bring so much value 🤷‍♂️

Comment (steebchen) [reply src/packages/migrate/src/utils/seed.ts:46]: I think it's more what it says than what it does; usually I use if-else if there are many different conditions which may be very different on what they do (like either throw or do some action, or create a database record or update something, etc.), but in this case it's the same action but a slightly different execution. It's more than a nit pick though so feel free to ignore me haha

File: src/packages/cli/src/bin.ts (hunk 1 above)
Comment (steebchen) [reply src/packages/cli/src/bin.ts]: @tomhoule you were too fast

File: src/packages/migrate/src/utils/seed.ts (hunk 2 above)
Comment (Jolg42) [reply src/packages/migrate/src/utils/seed.ts:26]: Done in a commit with small changes, Thanks!

File: src/packages/migrate/src/utils/seed.ts (hunk 3 above)
Comment (Jolg42) [reply src/packages/migrate/src/utils/seed.ts:45]: Thanks! Indeed I get the idea :)

File: src/packages/migrate/src/__tests__/DbSeed.test.ts [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  })
+
+  it('with missing db should fail', async () => {
+    ctx.fixture('seed-sqlite')
+    ctx.fs.remove('prisma/dev.db')
+    ctx.fs.remove('prisma/seed.ts')
+
+    const result = DbSeed.new().parse(['--preview-feature'])
+    await expect(result).rejects.toMatchInlineSnapshot(`
+            More than 1 seed file was found in \`prisma\`.
```
Comment (albertoperdomo) [inline src/packages/migrate/src/__tests__/DbSeed.test.ts:25]:
```suggestion
            More than one seed file was found in \`prisma\`.
```

File: src/packages/migrate/src/__tests__/DbSeed.test.ts [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            This command only supports 1 \`seed.ts\` or \`.js\` or \`.sh\` or \`.go\` file.
```
Comment (albertoperdomo) [inline src/packages/migrate/src/__tests__/DbSeed.test.ts:26]:
```suggestion
            This command only supports one seed file: Use \`seed.ts\`, \`.js\` or \`.sh\` or \`.go\`.
```

File: src/packages/migrate/src/__tests__/DbSeed.test.ts [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+      ctx.mocked['console.error'].mock.calls.join('\n'),
+    ).toMatchInlineSnapshot(``)
+  })
+
+  it('more than one seed file', async () => {
+    ctx.fixture('seed-sqlite')
+
+    const result = DbSeed.new().parse(['--preview-feature'])
+    await expect(result).rejects.toMatchInlineSnapshot(`
+            More than 1 seed file was found in \`prisma\`.
```
Comment (albertoperdomo) [inline src/packages/migrate/src/__tests__/DbSeed.test.ts:58]:
```suggestion
            More than one seed file was found in \`prisma\`.
```

File: src/packages/migrate/src/__tests__/DbSeed.test.ts [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+            This command only supports 1 \`seed.ts\` or \`.js\` or \`.sh\` or \`.go\` file.
```
Comment (albertoperdomo) [inline src/packages/migrate/src/__tests__/DbSeed.test.ts:59]:
```suggestion
            This command only supports one seed file: Use \`seed.ts\`, or \`.js\` or \`.sh\` or \`.go\`.
```

File: src/packages/migrate/src/__tests__/DbSeed.test.ts [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    ctx.fixture('seed-sqlite')
+    // ctx.fs.remove('prisma/seed.js')
+    ctx.fs.remove('prisma/seed.ts')
+    ctx.fs.remove('prisma/seed.sh')
+    ctx.fs.remove('prisma/seed.go')
+
+    const result = DbSeed.new().parse(['--preview-feature'])
+    await expect(result).resolves.toMatchInlineSnapshot(`
+
+                                                                                                                                                                                                            🌱  Your database is now seeded.
```
Comment (albertoperdomo) [inline src/packages/migrate/src/__tests__/DbSeed.test.ts:79]:
```suggestion
                                                                                                                                                                                                            🌱  Your database has been seeded.
```

File: src/packages/migrate/src/__tests__/DbSeed.test.ts [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+    ctx.fixture('seed-sqlite')
+    ctx.fs.remove('prisma/seed.js')
+    // ctx.fs.remove('prisma/seed.ts')
+    ctx.fs.remove('prisma/seed.sh')
+    ctx.fs.remove('prisma/seed.go')
+
+    const result = DbSeed.new().parse(['--preview-feature'])
+    await expect(result).resolves.toMatchInlineSnapshot(`
+
+                                                                                                                                                                                                            🌱  Your database is now seeded.
```
Comment (albertoperdomo) [inline src/packages/migrate/src/__tests__/DbSeed.test.ts:99]:
```suggestion
                                                                                                                                                                                                            🌱  Your database has been seeded.
```

File: src/packages/migrate/src/__tests__/DbSeed.test.ts [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+    ctx.fixture('seed-sqlite')
+    ctx.fs.remove('prisma/seed.js')
+    ctx.fs.remove('prisma/seed.ts')
+    // ctx.fs.remove('prisma/seed.sh')
+    ctx.fs.remove('prisma/seed.go')
+
+    const result = DbSeed.new().parse(['--preview-feature'])
+    await expect(result).resolves.toMatchInlineSnapshot(`
+
+                                                                                                                                                                                                            🌱  Your database is now seeded.
```
Comment (albertoperdomo) [inline src/packages/migrate/src/__tests__/DbSeed.test.ts:119]:
```suggestion
                                                                                                                                                                                                            🌱  Your database has been seeded.
```

File: src/packages/migrate/src/__tests__/fixtures/seed-sqlite/README.md [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+
+- `npm install`
+
+- Edit the `.env.example` with your PostgreSQL url
+
+- Rename the `.env.example` file to `.env`
+
+- Use `db push` to create the database structure `npx prisma db push --preview-feature`
+
+- Run the seed to populate the database `npm run seed`
```
Comment (albertoperdomo) [inline src/packages/migrate/src/__tests__/fixtures/seed-sqlite/README.md:11]: Should we not be using the prisma db seed command here?

File: src/packages/migrate/src/commands/DbSeed.ts [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+    if (!args['--preview-feature']) {
+      throw new PreviewFlagError()
+    }
+
+    await tryToRunSeed()
+
+    return `\n${
+      process.platform === 'win32' ? '' : '🌱  '
+    }Your database is now seeded.`
```
Comment (albertoperdomo) [inline src/packages/migrate/src/commands/DbSeed.ts:61]:
```suggestion
    }Your database has been seeded.`
```

File: src/packages/migrate/src/utils/seed.ts [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+export async function tryToRunSeed() {
+  const detected = detectSeedFiles()
+
+  if (detected.numberOfSeedFiles === 0) {
+    throw new Error(`No seed file found.
+Create a \`seed.ts\` or \`.js\` or \`.sh\` or \`.go\` file in the prisma folder.`)
+  } else if (detected.numberOfSeedFiles > 1) {
+    throw new Error(
+      `More than 1 seed file was found in \`${path.relative(
```
Comment (albertoperdomo) [inline src/packages/migrate/src/utils/seed.ts:38]:
```suggestion
      `More than one seed file was found in \`${path.relative(
```

File: src/packages/migrate/src/utils/seed.ts [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+        process.cwd(),
+        path.dirname(detected.seedPath),
+      )}\`.
+This command only supports 1 \`seed.ts\` or \`.js\` or \`.sh\` or \`.go\` file.`,
```
Comment (albertoperdomo) [inline src/packages/migrate/src/utils/seed.ts:42]:
```suggestion
This command only supports one seed file: User \`seed.ts\`, \`.js\` or \`.sh\` or \`.go\`.`,
```

**Review by albertoperdomo — APPROVED** [review]

**Review by albertoperdomo — CHANGES_REQUESTED** [review]
Looks fine but I would at least change 1 for one. It looks very odd IMO.
I would also change the phrasing.

Seeding feels like something that you do, you could also do it multiple times depending how you write your seeds, which is why I would rather express it as "has been seeded" (we just did it) vs it is seeded (feels binary , either seeded or not, but it does not feel like it could be ran multiple times).

File: src/packages/migrate/src/__tests__/fixtures/seed-sqlite/README.md (hunk 11 above)
Comment (Jolg42) [reply src/packages/migrate/src/__tests__/fixtures/seed-sqlite/README.md:11]: I forgot to remove that README actually, it's only for testing

File: src/packages/migrate/src/utils/seed.ts (hunk 3 above)
Comment (timsuchanek) [reply src/packages/migrate/src/utils/seed.ts:45]: `switch` is for noobs, you should use a map! 😄

**Comment by timsuchanek** [comment]
Looks good :)

**Comment by Jolg42** [comment]
It's in `2.15.0-dev.56` (For internal QA)

**Pull request merged**